examples/transport_wrapper.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright (c) 2020 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This example shows how to get the e-mail addresses of the creators of managed clusters.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/openshift-online/ocm-sdk-go"
+)
+
+type LoggingTransport struct {
+	logger    sdk.Logger
+	transport http.RoundTripper
+}
+
+func NewLoggingTransport(logger sdk.Logger) *LoggingTransport {
+	return &LoggingTransport{
+		logger: logger,
+	}
+}
+
+func (t *LoggingTransport) GetTransportWrapper(transport http.RoundTripper) http.RoundTripper {
+	t.transport = transport
```

Comment:
Not sure if this is a good example of how to do things. It makes the logging transport mutable. If you reuse it with multiple connectsions then all the requests will end up going to the last connection:

```go
// Create the logging transport:
loggingTransport := NewLoggingTransport(logger)

// Create two connections, using the same logging transport, with the
// intent to send the log to the same logger:
staging, err := sdk.NewConnectionBuilder().
        URL("https://staging").
        TransportWrapper(loggingTransport.GetTransportWrapper).
        Build()
if err != nil {
        ...
}
production, err := sdk.NewConnectionBuilder().
        URL("https://production").
        TransportWrapper(loggingTransport.GetTransportWrapper).
        Build()
if err != nil {
        ...
}

// Send a request apparently to "staging" but actually to "production":
_, err := staging.ClustersMgmt().V1().Clusters().Add().
        ....
        Send()
if err != nil {
        ...
}
```

I'd prefer to make the logging transport inmutable, and make the wrapper a function without side effects. Something like this:

```go
// NewLoggingTransport creates a transport that sends basic details of requests to the
// given logger. The wrapped transport will be used actually send the requests.
func NewLoggingTransport(logger sdk.Logger, wrapped http.RoundTripper) http.RoundTripper {
        return &LoggingTransport{
                logger: logger,
                wrapped: wrapped,
        }
}
```

When setting up the connection you can do then something like this:

```go
TransportWrapper(func (wrapped http.RoundTripper) http.RoundTripper {
        return NewLoggingTransport(logger, wrapped)
}).

Reply:
I think this creates a confusing example but I understand and agree with your point.
Fixed.